pkg/identityserver/contact_info_registry.go

```diff
@@ -106,17 +107,41 @@ func (is *IdentityServer) requestContactInfoValidation(
 	if err != nil {
 		return nil, err
 	}
+	now := time.Now()
+	ttl := is.configFromContext(ctx).UserRegistration.ContactInfoValidation.TokenTTL
+	expires := now.Add(ttl)
+	retryInterval := is.configFromContext(ctx).UserRegistration.ContactInfoValidation.RetryInterval
+
 	var contactInfo []*ttnpb.ContactInfo
+	resendValidation := make(map[string]*ttnpb.ContactInfoValidation)
+
 	err = is.store.Transact(ctx, func(ctx context.Context, st store.Store) (err error) {
 		contactInfo, err = st.GetContactInfo(ctx, ids)
-		return err
+		if err != nil {
+			return err
+		}
+
+		validations, err := st.ListValidations(ctx, ids)
+		if err != nil {
+			return err
+		}
+		for _, v := range validations {
+			if v.GetUpdatedAt().AsTime().Before(now.Add(-retryInterval)) {
```

Comment:
I would make something like `ListRefreshableValidations` to be honest - there is no need to filter here.